### PR TITLE
Add ngeo draw feature directive

### PR DESCRIPTION
**Pull request by adube**
This PR introduces a `ngeo-drawfeature` directive and an example featuring it.

## Todo

 * [x] deal with i18n values
 * [x] use contants for feature properties and geometry types once #955 is merged
 * [x] set the default style of the features when drawn once #955 is merged
 * [x] prevent the features and overlay that are added in the measure interactions to be shown, as those conflict with the actual drawn features.
 * [x] code review
 * [x] merge into one commit
 * [x] travis build passes

## Live example

 * http://adube.github.io/ngeo/edit-feature-draw/examples/drawfeature.html

### CODE REVIEW

**Comment by adube** [comment]
@fgravin The directive currently contains lots of copy/pasted code from many places: Luxembourg, ngeo examples, etc.  The i18n text contained within the directive are a result of that.  Would you please tell me what strategy I should use for the i18n ?

If you want to review what's in progress, go ahead.  Keep in mind what's remaining to do (see in the description).

**Comment by fgravin** [comment]
Looks good for what i've seen so far.

> Would you please tell me what strategy I should use for the i18n ?

Ok i will think about it.

**Comment by fgravin** [comment]
> The i18n text contained within the directive are a result of that. Would you please tell me what strategy I should use for the i18n

See https://github.com/Geoportail-Luxembourg/geoportailv3/blob/30272dc5426b6c2d60d7a9aa8d5d38843ca93360/geoportailv3/static/js/measure/measuredirective.js#L138

You need to pass some html to the constructor options.
This html must be compiled (`$compile`) with the directive `translate`, this way the string will be replaced on the fly if the language change.

**Comment by adube** [comment]
Understood.  I'm on it.

**Comment by adube** [comment]
@fgravin This is now ready for review.  The demo is up-to-date too.

File: contribs/gmf/src/directives/drawfeature.js [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+ * @ngInject
+ * @ngdoc directive
+ * @ngname gmfDrawfeature
+ */
+gmf.drawfeatureDirective = function() {
+  return {
+    controller: 'GmfDrawfeatureController',
+    scope: {
+      'active': '=gmfDrawfeatureActive',
+      'map': '=gmfDrawfeatureMap'
```
Comment (fgravin) [inline contribs/gmf/src/directives/drawfeature.js]: use `>` instead (one way binding)

File: contribs/gmf/src/directives/drawfeature.js [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+   */
+  this.drawText = new ol.interaction.Draw({
+    type: ol.geom.GeometryType.POINT
+  });
+
+  var drawText = this.drawText;
+  this.interactions_.push(drawText);
+  drawText.setActive(false);
+  ngeoDecorateInteraction(drawText);
+  this.map.addInteraction(drawText);
```
Comment (fgravin) [inline contribs/gmf/src/directives/drawfeature.js]:
```js
  interaction.setActive(false);
  ngeoDecorateInteraction(interaction);
  this.map.addInteraction(interaction);
```

maybe this could be factorized in a `this.interactions_.forEach()` ?
It's a bit heavy.

File: contribs/gmf/src/directives/drawfeature.js [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+      name = this.gettextCatalog_.getString('LineString');
+      break;
+    case ngeo.GeometryType.POLYGON:
+      name = this.gettextCatalog_.getString('Polygon');
+      break;
+    case ngeo.GeometryType.RECTANGLE:
+      name = this.gettextCatalog_.getString('Rectangle');
+      feature.set(prop.IS_RECTANGLE, true);
+      break;
+    default:
```
Comment (fgravin) [inline contribs/gmf/src/directives/drawfeature.js]:
```js
default:
  name = this.gettextCatalog_.getString(type)
```
and removing point, linestring, polygon cases ??

File: contribs/gmf/src/directives/drawfeature.js [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+};
+
+
+/**
+ * @param {ol.ObjectEvent} event Event.
+ * @private
+ */
+gmf.DrawfeatureController.prototype.handleActiveChange_ = function(event) {
+  var active;
+  this.interactions_.some(function(interaction) {
```
Comment (fgravin) [inline contribs/gmf/src/directives/drawfeature.js]: use the return value of the `some` function instead of a new `var active`.
Something like
```js
this.active = this.interactions_.some(function(interaction) {
  return interaction.getActive();
});
```

**Comment by fgravin** [comment]
Some minor comments, the overall looks good.
But i wonder if this shouldn't go into ngeo in a more customisable fashion.

File: contribs/gmf/src/directives/drawfeature.js (hunk 3 above)
Comment (adube) [inline contribs/gmf/src/directives/drawfeature.js]: I don't understand this comment.

Comment (adube) [inline contribs/gmf/src/directives/drawfeature.js]: Ah, okay I get it.  The constants currently have all lower case values.  I suppose it wouldn't hurt to make them camel case.  For `LineString`, that means that the constant should be renamed `LINE_STRING`, I suppose.

Comment (fgravin) [inline contribs/gmf/src/directives/drawfeature.js]: name = this.gettextCatalog_.getString('LineString');

This line could go in default ? As you are doing it for all

**Comment by adube** [comment]
Thanks for the review.  I'll address the changes required.

Please, let me know if this should go into ngeo instead.

**Comment by fgravin** [comment]
What i have thought of in the beginning was a directive draw like this, with the controller logic.
Then, one small directive per drawing type, requiring the draw directive, accessing its controller.

Each small directive register itself in the main controller, and adds its specific handlers.
Common handlers and whole logic would be in main draw directive, no template for this one.
It would go into ngeo.

What do you think ?

**Comment by adube** [comment]
@fgravin and I discussed about this.  It makes sense to move the draw directive in ngeo.  I'm on it.

**Comment by adube** [comment]
@fgravin Travis fails.  I don't understand why.

**Comment by sbrunner** [comment]
You see that the error is in the animation example ... https://travis-ci.org/camptocamp/ngeo/jobs/120962154#L546,, I also don't understand why it failed ...

**Comment by adube** [comment]
@fgravin Now, each draw/measure directive have its own file.

I still have the build issue when making `make check-examples` (see Travis build error).  I still don't know what's going on wrong.  That prevents me from updating the live example.  If you have a minute to check what's wrong I'd appreciate it.

**Comment by fgravin** [comment]
Do a `make hosted-example`and test your example in compiled mode (localhost hosted-examples ..).
You'll see an error, with a link, open the link if not to long, it can give help.

Otherwise, try to remove your subdirective from the template, see if you still have the error. Try to debug step by step by removing some code to see what's going on.
You don't miss any `@ngInject` ?

**Comment by adube** [comment]
@fgravin Thanks for the tip.  All other examples but mine were throwing an error.  That was because of the addition of `gettext` as dependency in ngeo.  Now, all examples properly include the .js file needed.

With this, we're good to go.

**Comment by adube** [comment]
Live example updated.

File: src/directives/drawfeature.js [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+   * @type {Array.<ol.interaction.Interaction>}
+   * @private
+   */
+  this.interactions_ = [];
+
+  /**
+   * @type {ol.interaction.Draw}
+   * @export
+   */
+  this.drawPoint;
```
Comment (fgravin) [inline src/directives/drawfeature.js]: I think we should remove it from here.
Cause the main directive doesn't know what interaction it can have.
Do you agree ?

Comment (adube) [inline src/directives/drawfeature.js]: Mmm... I liked having those listed here.  Those 6 draw/measure directives can only be used within this main drawFeature directive, and those properties identify them correctly.  If one looks at this file only, one will understand the different interactions that can be bound from here.

I'm not against removing them.  I just find it useful.  Also, I haven't tested but this makes the properties correctly exported by closure, I think.  Otherwise, we would probably need to do `drawFeatureCtrl['drawPoint'] = drawPoint`.

Bottom line: I'm in favour of keeping them.  What do you think ?

**Comment by fgravin** [comment]
Can you rebase please ?

**Comment by adube** [comment]
Done.

File: src/directives/measureazimut.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    require: '^^ngeoDrawfeature',
+    /**
+     * @param {!angular.Scope} $scope Scope.
+     * @param {angular.JQLite} element Element.
+     * @param {angular.Attributes} attrs Attributes.
+     * @param {ngeo.DrawfeatureController} drawFeatureCtrl Controller.
+     */
+    link: function($scope, element, attrs, drawFeatureCtrl) {
+
+      var helpMsg = gettext('Click to start drawing azimut');
```
Comment (fgravin) [inline src/directives/measureazimut.js]: I don't really get it, i know it comes from Lux.
What is `helpMsg` value, the translation of the key ?

Does `translate` directive below is usefull ? I think something is redundant here.

Comment (adube) [inline src/directives/measureazimut.js]: You tell me.

File: contribs/gmf/src/directives/partials/search.html [hunk 7]
```diff
@@ -8,4 +8,5 @@
 <span class="clear-button ng-hide"
         ng-hide="!ctrl.clearButton || ctrl.input_value == ''"
         ng-click="ctrl.onClearButton()"></span>
+  <input data-ng-model="color" data-ng-change="ctrl.setStyleColor(color)"></input>
```
Comment (ger-benjamin) [inline contribs/gmf/src/directives/partials/search.html]: @adube Can you explain me what is that ? That creates a field below the search field that seems not used.
Is that an error ? What's the aim ?

Comment (pgiraud) [inline contribs/gmf/src/directives/partials/search.html]: Maybe related to this https://github.com/camptocamp/c2cgeoportal/wiki/Specs-%23869---Color-of-search-result

Comment (adube) [inline contribs/gmf/src/directives/partials/search.html]: This was part of a patch that @fgravin did for me.  I hadn't noticed it, but it shouldn't have been added.  You can remove this.

Comment (ger-benjamin) [inline contribs/gmf/src/directives/partials/search.html]: Thanks for these information, I will remove that temporarily.
